lib/graphql/execution/interpreter/runtime.rb

```diff
@@ -304,43 +321,87 @@ def evaluate_selection_with_args(kwarg_arguments, field_defn, next_path, ast_nod
+                (nn = selection_result.graphql_non_null) &&
+                (nn == true || nn.include?(result_name))
+              # This is an invalid nil that should be propagated
+              parent = selection_result.graphql_parent
+              name_in_parent = selection_result.graphql_result_name
+              if parent.nil? # This is a top-level result hash
+                @response = nil
+              else
+                set_result(parent, name_in_parent, nil)
+              end
+            else
+              selection_result[result_name] = value
+            end
+          end
+        end
+
         HALT = Object.new
-        def continue_value(path, value, parent_type, field, is_non_null, ast_node)
-          if value.nil?
+        def continue_value(path, value, parent_type, field, is_non_null, ast_node, result_name, selection_result) # rubocop:disable Metrics/ParameterLists
+          case value
+          when nil
             if is_non_null
               err = parent_type::InvalidNullError.new(parent_type, field, value)
-              write_invalid_null_in_response(path, err)
+              if !selection_result.graphql_dead
+                schema.type_error(err, context)
+                set_result(selection_result, result_name, nil)
+                selection_result.graphql_dead = true
+              end
             else
-              write_in_response(path, nil)
+              set_result(selection_result, result_name, nil)
             end
             HALT
-          elsif value.is_a?(GraphQL::ExecutionError)
-            value.path ||= path
-            value.ast_node ||= ast_node
-            write_execution_errors_in_response(path, [value])
-            HALT
-          elsif value.is_a?(Array) && value.any? && value.all? { |v| v.is_a?(GraphQL::ExecutionError) }
-            value.each_with_index do |error, index|
-              error.ast_node ||= ast_node
-              error.path ||= path + (field.type.list? ? [index] : [])
+          when GraphQL::Error
+            # Handle these cases inside a single `when`
+            # to avoid the overhead of checking three different classes
+            # every time.
+            if value.is_a?(GraphQL::ExecutionError)
+              if !selection_result.graphql_dead
+                value.path ||= path
+                value.ast_node ||= ast_node
+                context.errors << value
+                set_result(selection_result, result_name, nil)
+              end
+              HALT
+            elsif value.is_a?(GraphQL::UnauthorizedError)
+              # this hook might raise & crash, or it might return
+              # a replacement value
+              next_value = begin
+                schema.unauthorized_object(value)
+              rescue GraphQL::ExecutionError => err
+                err
+              end
+              continue_value(path, next_value, parent_type, field, is_non_null, ast_node, result_name, selection_result)
+              HALT
```

Comment:
Oops, this is a bug. Previously it didn't return `HALT`, it returned `continue_value(...)`. I should remove this `HALT`.